File: 2. Add Two Numbers/solution.cpp
/*
  Runtime: 20 ms, faster than 99.22% of C++ online submissions for Add Two Numbers.
  Memory Usage: 13.7 MB, less than 56.62% of C++ online submissions for Add Two Numbers.
*/

void GiveResult(ListNode* res, ListNode* l1, ListNode* l2, int x){
    int sum = x;
    l1 ? sum += l1->val : sum += 0;
    l2 ? sum += l2->val : sum += 0;
    if(!l1 && !l2 && x == 0)
        return;
    ListNode* tmp = new ListNode(1);
    res->next = tmp;
    if (!l1 && !l2)
        return;
    l1 ? l1 = l1->next : l1;
    l2 ? l2 = l2->next : l2;

    if (sum > 9){
        res->next->val = sum - 10;
        GiveResult(res->next, l1, l2, 1);
    } else {
        res->next->val = sum;
        GiveResult(res->next, l1, l2, 0);
    }
}
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode* res = new ListNode(9);
        GiveResult(res, l1, l2, 0);
        return res->next;
    }
};

File: 2. Add Two Numbers/solution.py
#   Runtime: 72 ms, faster than 98.68% of Python3 online submissions for Add Two Numbers.
#   Memory Usage: 13.5 MB, less than 5.36% of Python3 online submissions for Add Two Numbers.

def create_result(res, remainder, l1=None, l2=None):
    sum = remainder
    if l1: sum += l1.val
    if l2: sum += l2.val
    if not l1 and not l2 and remainder == 0:
        return
    res.next = ListNode(1)
    if not l1 and not l2:
        return
    l1 = l1.next if l1 else l1
    l2 = l2.next if l2 else l2
    if sum > 9:
        res.next.val = sum - 10
        create_result(res.next, 1, l1, l2);
    else:
        res.next.val = sum;
        create_result(res.next, 0, l1, l2);
        
class Solution(object):
    def addTwoNumbers(self, l1, l2):
        res = ListNode(1)
        create_result(res, 0, l1, l2);
        return res.next
